File: hrm_backend/src/models/employee.model.js
const mongoose = require("mongoose");
const Schema = mongoose.Schema;

const employeeSchema = new Schema(
  {
    user: {
      type: Schema.Types.ObjectId,
      ref: "User",
      required: true,
      unique: true,
    },
    email: {
      type: String,
      required: true,
      unique: true,
      trim: true,
      lowercase: true,
    },
    name: {
      type: String,
      trim: true,
      default: "",
    },
    subOrganisation: {
      type: String,
      required: true,
      trim: true,
    },
    employeeId: {
      type: String,
      required: true,
      unique: true,
      trim: true,
    },
    profilePic: {
      type: String,
      default: "",
    },
    phone: {
      type: String,
      default: "",
    },
    jobTitle: {
      type: String,
      default: "",
    },
    bloodGroup: {
      type: String,
      default: "",
    },
    joinDate: {
      type: Date,
      default: null,
    },
    // EOD Warning System
    eodWarningCount: {
      type: Number,
      default: 0,
      min: 0,
    },
    payCutFlag: {
      type: Boolean,
      default: false,
    },
    // Leave Management System
    leaveBalance: {
      casualLeave: {
        type: Number,
        default: 2, // 2 per month, carries over
      },
      workFromHome: {
        type: Number,
        default: 1, // 1 per month, resets monthly
      },
      permissionHours: {
        type: Number,
        default: 3, // 3 hours per month, resets monthly
      },
    },
    leaveHistory: {
      lastResetDate: {
        type: Date,
        default: Date.now,
      },
      totalCasualLeavesUsed: {
        type: Number,
        default: 0,
      },
      totalSickLeavesUsed: {
        type: Number,
        default: 0,
      },
      totalWFHUsed: {
        type: Number,
        default: 0,
      },
      totalPermissionHoursUsed: {
        type: Number,
        default: 0,
      },
    },
    // Messaging Permissions System
    messagingPermissions: {
      canMessage: {
        type: Boolean,
        default: false,
      },
      grantedBy: {
        type: Schema.Types.ObjectId,
        ref: "Admin",
        default: null,
      },
      grantedAt: {
        type: Date,
        default: null,
      },
      expiresAt: {
        type: Date,
        default: null,
      },
      lastRequestedAt: {
        type: Date,
        default: null,
      },
      requestedFrom: {
        type: Schema.Types.ObjectId,
        ref: "Admin",
        default: null,
      },
    },
  },
  {
    timestamps: true,
  }
);

const Employee = mongoose.model("Employee", employeeSchema);

module.exports = Employee;

File: hrm_app/lib/data/models/employee_model.dart
import 'package:hrm_app/data/models/user_model.dart';

class Employee {
  final String id;
  final String name;
  final String email;
  final String phone;
  final String profilePic;
  final String jobTitle;
  final String subOrganisation;
  final String employeeId;
  final String? bloodGroup;
  final UserModel? user;

  Employee({
    required this.id,
    required this.name,
    required this.email,
    required this.phone,
    required this.profilePic,
    required this.jobTitle,
    required this.subOrganisation,
    required this.employeeId,
    this.bloodGroup,
    this.user,
  });

  factory Employee.fromJson(Map<String, dynamic> json) {
    return Employee(
      id: json['_id'] as String,
      name: json['name'] as String? ?? 'No Name',
      email: json['email'] as String? ?? 'No Email',
      phone: json['phone'] as String? ?? '',
      profilePic: json['profilePic'] as String? ?? '',
      jobTitle: json['jobTitle'] as String? ?? '',
      subOrganisation: json['subOrganisation'] as String? ?? '',
      employeeId: json['employeeId'] as String? ?? '',
      bloodGroup: json['bloodGroup'] as String?,
      user: json['user'] != null
          ? UserModel.fromJson(json['user'] as Map<String, dynamic>)
          : null,
    );
  }
}
